core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,424 @@
+                .last_miniblock_number(ctx)
+                .await
+                .wrap("last_miniblock_number()")?;
+            if number >= block.header().number {
+                storage
+                    .insert_certificate(ctx, &block.justification, self.inner.operator_address)
+                    .await
+                    .wrap("insert_certificate()")?;
+                return Ok(());
+            }
+            drop(storage);
+            ctx.sleep(POLL_INTERVAL).await?;
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl ReplicaStore for Store {
+    async fn state(&self, ctx: &ctx::Ctx) -> ctx::Result<Option<ReplicaState>> {
+        let storage = &mut CtxStorage::access(ctx, &self.pool).await.wrap("access()")?;
+        storage.replica_state(ctx).await.wrap("replica_state()")
+    }
+
+    async fn set_state(&self, ctx: &ctx::Ctx, state: &ReplicaState) -> ctx::Result<()> {
+        let storage = &mut CtxStorage::access(ctx, &self.pool).await.wrap("access()")?;
+        storage
+            .set_replica_state(ctx, state)
+            .await
+            .wrap("set_replica_state()")
+    }
+}
+
+#[async_trait::async_trait]
+impl PayloadManager for Store {
+    /// Currently (for the main node) proposing is implemented as just converting a miniblock from db (without a cert) into a
+    /// payload.
+    async fn propose(
+        &self,
+        ctx: &ctx::Ctx,
+        block_number: validator::BlockNumber,
+    ) -> ctx::Result<validator::Payload> {
+        const POLL_INTERVAL: time::Duration = time::Duration::milliseconds(50);
+        let mut storage = CtxStorage::access(ctx, &self.pool).await.wrap("access()")?;
+        storage
+            .certificate(ctx, block_number.prev())
+            .await
+            .wrap("certificate()")?
+            .with_context(|| format!("parent of {block_number:?} is missing"))?;
+        drop(storage);
+        loop {
+            let mut storage = CtxStorage::access(ctx, &self.pool).await.wrap("access()")?;
+            if let Some(payload) = storage
+                .payload(ctx, block_number, self.operator_address)
+                .await
+                .wrap("payload()")?
+            {
+                return Ok(payload.encode());
+            }
+            drop(storage);
+            ctx.sleep(POLL_INTERVAL).await?;
+        }
+    }
+
+    /// Verify that `payload` is a correct proposal for the block `block_number`.
+    /// Currently (for the main node) it is implemented as checking whether the received payload
+    /// matches the miniblock in the db.
+    async fn verify(
+        &self,
+        ctx: &ctx::Ctx,
+        block_number: validator::BlockNumber,
+        payload: &validator::Payload,
+    ) -> ctx::Result<()> {
+        let want = self.propose(ctx, block_number).await?;
+        let want = consensus::Payload::decode(&want).context("Payload::decode(want)")?;
+        let got = consensus::Payload::decode(payload).context("Payload::decode(got)")?;
```

Comment:
Nit: `consensus::` qualification is redundant; `Payload` is imported into the scope directly.